### DISCUSSION TITLE
Could a Composer plugin modify the version range used by composer require?

### DISCUSSION THREAD

**Discussion by mxr576** [composer/composer#11373] [Q&A / Support] [answered]
I am curious if there is a way to modify the version (version range) via a plugin that gets added to `composer.json` when a new dependency is installed by `composer require`. Is there an event that would allow to make a change like this?

My goal is to enforce a specific version range (like `~3.0.0` instead of `^3.0.0`) or even lock on a version (`3.1.1` instead of `^3.1.0`) when a new component is installed by `composer require.`

**Comment by Seldaek** [comment, marked as answer]
For the latter you can use `--fixed` when calling require but there's no way to enforce this. Well a plugin I guess could listen to the `pre-command-run` event and modify the input to force --fixed to always be set.. But there's no way otherwise to hook into the version resolution. 

I don't really see a way to do this sorry.

**Reply by mxr576** [reply]
> Well a plugin I guess could listen to the pre-command-run event and modify the input to force --fixed to always be set..

Interesting workaround for " lock on a version (3.1.1 instead of ^3.1.0) when a new component is installed by" - could work only the list only contains packages what I would like enforce this rule, if there another package on the list, most probably the plugin should just abort the process and ask the user to normalize the input.

>  enforce a specific version range (like ~3.0.0 instead of ^3.0.0)

So there is no way to enforce usage of `~` by default (like when no version is specified or override `^x.y.z` to `~x.y.z`) form a plugin?
Enforcing `~` could be useful for those packages that implements semver a "different way".

**Reply by Seldaek** [reply]
Yeah no I don't see a way. We never use `~` because really people should follow semver ideally, and always using `^` makes for less confusion.

**Reply by mxr576** [reply]
Okay, thanks for the answer! :)
